account_invoice_merge/models/account_invoice.py

```diff
@@ -4,214 +4,180 @@
         qty_prec = self.env['decimal.precision'].precision_get(
-            'Product Unit of Measure')
-        for invoice_key, (invoice_data, old_ids) in new_invoices.iteritems():
-            # skip merges with only one invoice
-            if len(old_ids) < 2:
-                allinvoices += (old_ids or [])
+            'Product Unit of Measure'
+        )
+        # we only work with draft invoices. We have a dirty check for
+        # this in wizard as well
+        draft_invoices = self.filtered(lambda x: x.state == 'draft')
+        grouped_invoices = draft_invoices.group_by_fields()
+        result = {}
+        for group in grouped_invoices:
+            # Ignore groups of 1 invoice as there is nothing to merge there
+            if len(group) < 2:
                 continue
-            # cleanup invoice line data
-            for key, value in invoice_data['invoice_line_ids'].iteritems():
-                value.update(dict(key))
-
-            if remove_empty_invoice_lines:
-                invoice_data['invoice_line_ids'] = [
-                    (0, 0, value) for value in
-                    invoice_data['invoice_line_ids'].itervalues() if
-                    not float_is_zero(
-                        value['quantity'], precision_digits=qty_prec)]
-            else:
-                invoice_data['invoice_line_ids'] = [
-                    (0, 0, value) for value in
-                    invoice_data['invoice_line_ids'].itervalues()]
-
+            # Building values for the new invoice
+            # As this values are the same across the group
+            # we can simply take them from the first invoice of the group
+            vals = group[0].get_group_values()
+            if keep_references:
+                vals['name'] = " ".join(
+                    (x for x in group.mapped("name") if x)
+                )
             if date_invoice:
-                invoice_data['date_invoice'] = date_invoice
-
-            # create the new invoice
-            newinvoice = self.with_context(is_merge=True).create(invoice_data)
-            invoices_info.update({newinvoice.id: old_ids})
-            allinvoices.append(newinvoice.id)
-            allnewinvoices.append(newinvoice)
+                vals["date_invoice"] = date_invoice
+            vals.update({
+                'origin': " ".join(
+                    (x for x in group.mapped("origin") if x)
+                ),
+                'reference': " ".join(
+                    (x for x in group.mapped("reference") if x)
+                ),
+            })
+            lines = []
+            # merge invoice lines
+            line_ids = group.mapped("invoice_line_ids")
+            for line_group in line_ids.group_by_fields():
+                line_vals = line_group[0].get_group_values()
+                line_vals['quantity'] = sum(line_group.mapped('quantity'))
+                if remove_empty_invoice_lines:
+                    # TODO: not sure this check is needed as
```

Comment:
Did you check it?
If you create an invoice with invoice lines, can you later modify them with a `0` quantity?